### ISSUE TITLE
make bare-bones readers MDAnalysis-independent

### ISSUE THREAD

**Issue by orbeckst** [MDAnalysis/mdanalysis#1194]
**In the long run** it would be useful to move forward with making trajectory readers MDAnalysis-independent. @kain88-de  has already done this for the Gromacs XDR reader. What I mean by this is that someone else could, in principle, use `lib.formats` and then use it independently from MDA itself.

Why?

Because between MDAnalysis, mdtraj, pytraj, VMD … we are duplicating code and that's a waste of time and bad software engineering. It would be much better if (eventually) we (the community) have bare-bones readers and then everyone else can use them as part of their library. Generally, people liked that idea, judging from my interactions with others at meetings. Certainly, something like this is not happening overnight but we could take a leadership role and start by moving some of our readers/writers into `lib.format`. We would then be ready to move forward if we wanted to.

A [GSOC project](https://github.com/MDAnalysis/mdanalysis/wiki/Google-Summer-Of-Code) related to trajectory formats could be to

1. formally define the api (basically, numpy arrays or array_like)… follow the Gromacs XDR code
2. move some/all readers/writers to lib.format 

See also https://groups.google.com/d/msg/mdnalysis-devel/QlwoL1_Etns/cqAZRrp0FwAJ

**Comment by dotsdl** [comment]
I'm in favor of this. If we can at least get to the point where our Readers (and perhaps eventually Writers) are clearly decoupled from the rest of the library then we can make eventual splitting out easy if it's of interest to other packages. MDAnalysis then can remain as a library focused on a nice user interface as it always has.

**Comment by dotsdl** [comment]
Can help to reduce the maintenance burden of an always-increasing number of Readers on the MDAnalysis devs alone, too.

**Comment by kain88-de** [comment]
Btw this means that we have to be careful with contributions to these code-pieces now. If we want to have a separate library it can't have a copy-left license like the GPL, a BSD license would be more appropriate. We can only do this if we keep close tabs on the contributions.

Actually what we should do then is after the 0.16. release is to create a new python project that starts with the xtc format and give it a liberal license like BSD. It would be a new dependency of MDAnalysis but that would be the case eventually anyway.

**Comment by orbeckst** [comment]
Regarding licensing: the library could certainly have a GPL – that's a choice. It is true that this could potentially decrease adoption but consider that, for instance, mdtraj is LGPL. 

Until we have a spirited discussion about this I agree with you: we need to keep good track of commiters that we could ask for a relicence if necessary.

**Comment by jbarnoud** [comment]
I do not care much about the license if it is open source. I know @mnmelo had opinions on the question, though.

**Comment by kain88-de** [comment]
@orbeckst This would mean that only core developers can contribute to that part of the code right now. We have to be specific about that, otherwise it might be hard to track down people.

I checked the license, the LGPL will be ok for most projects since they can just separately install the library in python and load it in the python interpreter. This should also work for byte-compiled projects if they install the library with the original sources. Can be done and it isn't as much a pain point as for other compiled languages. We should also be ok with most MD formats that we have to link against in that case. The gromacs libraries are BSD licensed and for the others we have written pure python wrappers anyway.

**Comment by mnmelo** [comment]
I think the FSF page ["Why you shouldn't use the Lesser GPL for your next library"](https://www.gnu.org/licenses/why-not-lgpl.en.html) neatly summarizes my stance on this matter. It clarifies the common attitude that is "If it's a library then it should be LGPL".

In my view the readers are no different from MDAnalysis' core functionality, and if the latter is only available to copyleft users, why shouldn't the former? What we have to consider here is whether making this GPL will hinder development either due to lack of developers or due to lack of usage (namely, integration with mdtraj and others). My opinion is still to go for GPL, but this, of course, is a decision to be taken as a team.

**Comment by kain88-de** [comment]
> namely, integration with mdtraj and others

mdtraj for example is licensed with a LGPL.  They can't use a GPL code without re-licensing all of their code. I'm fairly certain they don't want to do that, as to other projects like pyemma (LGPL) who depend on it.

So a LGPL seems more appropriate for such a module.

**Comment by orbeckst** [comment]
Yeah, probably LGPL.

**Comment by orbeckst** [comment]
A step in the right direction: [MDAnalysis/pytng](https://github.com/MDAnalysis/pytng), see also #1366

**Comment by kain88-de** [comment]
We talked about the readers today at the dev meeting. Our Conclusion was to split out the readers into separate packages similar to pytng. I would start with libmdaxdr once the dcd reader is done

**Comment by orbeckst** [comment]
What is the rationale for separate packages as opposed to say, one package, with all the readers?

Did you talk about licenses?

**Comment by kain88-de** [comment]
This was originally @richardjgowers idea. But I also came to like it.  Here we can choose appropriate licenses per package. We can quickly push fixes for a single reader without updating a large dependency. Other projects can just take the readers they need instead of all.

**Comment by richardjgowers** [comment]
@orbeckst I preferred smaller packages as they often have weird dependencies (netcdf/hdf5/tng/etc) so if we only made them available bundled together it would result in quite a heavy to install package.  We could look at also packaging them together as a single package of packages though.

**Comment by orbeckst** [comment]
I see the advantages of small packages as you outline. But I think it makes it harder to ensure a consistent API, it is a (small) hassle to have 15 different mini packages as dependencies, you might duplicate testing code, and I think one would gain more recognition with a universal low-level reader library than an assortment of packages. 

I am also very much of the opinion that there should only be _one_ license for the readers – having readers with different licenses would be an interoperability disaster. Licensing needs to be figured out sooner than later and not postponed.

**Comment by Luthaf** [comment]
Hello !

I am working on a library that does exactly this: [chemfiles](http://chemfiles.org/) ([github repo](https://github.com/chemfiles/chemfiles/), [python API](http://chemfiles.org/chemfiles.py/latest/)).

The basic idea behind chemfiles is to prevent duplication of this kind reader/writer code, and to provide a single interface to [all supported formats](http://chemfiles.org/chemfiles/latest/formats.html#list-of-supported-formats). The core code is written in modern C++, with a C API and from there multiple languages interfaces, including Python 2 and 3. It is BSD licensed (for the next release, it was previously under MPLv2), and available on [conda-forge](https://github.com/conda-forge/chemfiles-python-feedstock/).

Looking though the issues on this repo, chemfiles could also fix #1133, #865 and #55.

I can see a few drawbacks with respect to the usage of chemfiles in MDAnalysis: 

1. It is C++ code, which means one needs to compile it before being able to use it. This is OK when using the conda distribution, but I don't have any PyPi wheel.
2. I am not working with biomolecules but with in materials science. This means that the API and data storage is not tailored for the specificity of biomolecules (only residues are supported, no chain, no tertiary structure, ...).

The good new on point 2. is that I would like to improve this part. I just don't know how biomolecules are usually structured, and I never had the time to think about it. The library is currently in version 0.7 (I will soon release 0.8), so breaking changes are definitively possible.

Please tell me what you think!

**Comment by kain88-de** [comment]
Hi, thanks for showing your library. It looks interesting.

Unfortunately not being able to install via pip is a huge downside for us. Such a reader as yours would be a core dependency and we try to be easily installable for a normal user (even without conda).

About your second point. Reading the topology is not part of this idea. If we were to include that we would need to include major parts of MDAnalysis (almost everything in core). For topologies, it might make more sense to have more specialized codes to handle the different needs.

**Comment by mnmelo** [comment]
Compiling via setup.py is very feasible, and we do it ourselves when doing wheel-free installs.

**Comment by richardjgowers** [comment]
Hello!

I'm imagining that the readers will be much faster than our pure Python ones, but I'm curious if **all** the corner cases of the annoying formats (like PDB) will work.  I guess it'd be fun to pick some files from our testsuite and see what the difference in performance (and error handling) is.

**Comment by Luthaf** [comment]
> Unfortunately not being able to install via pip is a huge downside for us. Such a reader as yours would be a core dependency and we try to be easily installable for a normal user (even without conda).

I think I could easily create wheels for chemfiles, and they would be universal wheels (no dependency on Python version), just by taking the conda build and re-packaging it. I'll give this a look.

> About your second point. Reading the topology is not part of this idea. If we were to include that we would need to include major parts of MDAnalysis (almost everything in core). For topologies, it might make more sense to have more specialized codes to handle the different needs.

Ok. So your use case would only be reading positions/velocity, right?

> I'm curious if all the corner cases of the annoying formats (like PDB) will work. I guess it'd be fun to pick some files from our testsuite and see what the difference in performance (and error handling) is.

I am certain all corner cases are not handled the same way (see https://github.com/chemfiles/chemfiles/issues/132 😄), but this can be resolved. Running chemfiles on MDAnalysis test suite could reveal more of these differences. The test files are all the files in `testsuite/MDAnalysisTests/data`, right?

**Comment by kain88-de** [comment]
> I think I could easily create wheels for chemfiles, and they would be universal wheels (no dependency on Python version), just by taking the conda build and re-packaging it. I'll give this a look.

We used to build our own conda package. A problem that wheels face as well is that you need to compile against an old version of libc. Otherwise, it's not gonna be compatible with other Linux distros. The manylinux project has a docker image that should be helpful. I don't have any tips for osx builds though.

**Comment by Luthaf** [comment]
I think conda-forge is already managing the [build against old libc](https://github.com/conda-forge/docker-images/blob/master/linux-anvil/Dockerfile#L1) and [set macOS minimul version](https://github.com/chemfiles/chemfiles-python-feedstock/blob/32504ee2b0b5da0f8f1d3e282c7f5d39bbd8ebfe/.travis.yml#L7) hard part of the work.

I might need to triple check that libstdc++ is also adequately set, or just link it statically anyway.

**Comment by kain88-de** [comment]
yeah but does conda-forge help to build a wheel?

**Comment by Luthaf** [comment]
No, but my idea was to take the compiled version from conda-forge, and repackage it in the wheel format.

**Comment by Luthaf** [comment]
> No, but my idea was to take the compiled version from conda-forge, and repackage it in the wheel format.

I got this working here: https://github.com/chemfiles/conda-to-wheels. You can download the wheels for testing [here](https://github.com/chemfiles/conda-to-wheels/releases/tag/0.7.4-rc) =). 

This was actually easier that I expected, knowing how Python packaging can be a pain sometimes!

**Comment by Luthaf** [comment]
Hi there! I just release version 0.8 of chemfiles, with real wheels: https://pypi.org/project/chemfiles/.

Linux wheels are built with the manylinux docker image, windows wheels are cross-compiled using mingw64 and statically link the libstdc++, and OS X wheels are built by setting MACOS_DEPLOYMENT_TARGET to 10.9. They all should work everywhere, but you can test them locally by running 

```
pip install chemfiles
python -c "from chemfiles import Atom; print(Atom('Fe').mass())"
```

If you are still interested in including chemfiles with MDAnalysis, what would be the steps forward? If you point me to the places to modify to integrate a new reader, I might have some time in the next weeks to try to get it to work and send a PR.

**Comment by kain88-de** [comment]
You can create a new reader with the format `CHEMFILES`.  The dcd reader should work as a good example. To use the reader one can then do `mda.Universe(top, traj, format='CHEMFILES')`.
